### PR TITLE
!str #6577 Align API of Flexi with Stage

### PR DESCRIPTION
**Pull request by patriknw**
- error -> fail
- complete -> finish
- onComplete - onUpstreamFinish
- onError - onUpstreamFailure
- onCancel - onDownstreamFinish

### CODE REVIEW

File: akka-docs-dev/rst/scala/code/docs/stream/FlexiDocSpec.scala [hunk 1]
```diff
@@ -127,10 +127,10 @@ class FlexiDocSpec extends AkkaSpec {
 
         override def initialCompletionHandling =
           CompletionHandling(
-            onComplete = (ctx, input) => input match {
+            onUpstreamFinish = (ctx, input) => input match {
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/stream/FlexiDocSpec.scala]: esp. here in CompletionHandling this rename helps a lot :+1:

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala [hunk 2]
```diff
@@ -109,7 +109,7 @@ object GraphFlexiMergeSpec {
 
 class TripleCancellingZip[A, B, C](var cancelAfter: Int = Int.MaxValue) extends FlexiMerge[(A, B, C)] {
   import FlexiMerge._
-  val soonCancelledInput = createInputPort[A]()
+  val soonDownstreamFinishledInput = createInputPort[A]()
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala]: typo: `s/Finishled/Finished/g`

File: akka-stream/src/main/scala/akka/stream/javadsl/FlexiMerge.scala [hunk 3]
```diff
@@ -219,31 +219,31 @@ object FlexiMerge {
...
-          ctx.error(cause)
+        override def onUpstreamFailure(ctx: MergeLogicContext[Out], input: InputHandle, cause: Throwable): State[A, Out] = {
+          ctx.fail(cause)
           sameState
         }
       }
 
     /**
-     * Completes as soon as any input completes.
+     * Completes as soon as any input finishs.
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/FlexiMerge.scala]: minor typo: `finishes`

File: akka-stream/src/main/scala/akka/stream/javadsl/FlexiMerge.scala [hunk 4]
```diff
@@ -219,31 +219,31 @@ object FlexiMerge {
     }
 
     /**
-     * Will continue to operate until a read becomes unsatisfiable, then it completes.
+     * Will continue to operate until a read becomes unsatisfiable, then it finishs.
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/FlexiMerge.scala]: `s/finishs/finishes/`

File: akka-stream/src/main/scala/akka/stream/scaladsl/FlexiMerge.scala [hunk 5]
```diff
@@ -184,37 +184,37 @@ object FlexiMerge {
...
      */
     sealed case class CompletionHandling(
-      onComplete: (MergeLogicContext, InputHandle) ⇒ State[_],
-      onError: (MergeLogicContext, InputHandle, Throwable) ⇒ State[_])
+      onUpstreamFinish: (MergeLogicContext, InputHandle) ⇒ State[_],
+      onUpstreamFailure: (MergeLogicContext, InputHandle, Throwable) ⇒ State[_])
 
     /**
-     * Will continue to operate until a read becomes unsatisfiable, then it completes.
+     * Will continue to operate until a read becomes unsatisfiable, then it finishs.
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlexiMerge.scala]: finishes

File: akka-stream/src/main/scala/akka/stream/scaladsl/FlexiMerge.scala [hunk 6]
```diff
@@ -184,37 +184,37 @@ object FlexiMerge {
...
      */
     val defaultCompletionHandling: CompletionHandling = CompletionHandling(
-      onComplete = (_, _) ⇒ SameState,
-      onError = (ctx, _, cause) ⇒ { ctx.error(cause); SameState })
+      onUpstreamFinish = (_, _) ⇒ SameState,
+      onUpstreamFailure = (ctx, _, cause) ⇒ { ctx.fail(cause); SameState })
 
     /**
-     * Completes as soon as any input completes.
+     * Completes as soon as any input finishs.
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlexiMerge.scala]: last one: finishes

**Comment by ktoso** [comment]
After typos fixed LGTM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala (hunk 2 above)
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala]: ha, `onCancel` search replace, thanks

File: akka-stream/src/main/scala/akka/stream/scaladsl/FlexiMerge.scala (hunk 5 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlexiMerge.scala]: thanks, I was too aggressive/ignorant in my replacements

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala [hunk 7]
```diff
@@ -546,16 +546,16 @@ class GraphFlexiMergeSpec extends AkkaSpec {
       s.expectNext("onInput: e")
       s.expectNext("onInput: c")
       s.expectNext("onInput: f")
-      s.expectNext("onComplete: 2")
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala]: not sure here. Do we test low level stuff so RS terminology makes sense here?

File: akka-stream/src/main/scala/akka/stream/impl/FlexiMergeImpl.scala [hunk 8]
```diff
@@ -63,13 +63,13 @@ private[akka] class FlexiMergeImpl(_settings: MaterializerSettings,
...
 
-    override def complete(): Unit = {
+    override def finish(): Unit = {
       inputBunch.cancel()
       primaryOutputs.complete()
       context.stop(self)
     }
 
-    override def error(cause: Throwable): Unit = fail(cause)
+    override def fail(cause: Throwable): Unit = FlexiMergeImpl.this.fail(cause)
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/FlexiMergeImpl.scala]: good

**Comment by drewhk** [comment]
LGTM, very good

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala (hunk 7 above)
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiMergeSpec.scala]: the expected string `"onUpstreamFinish` is generated by the TestMerge at `CompletionHandling.onUpstreamFinish`, so for me it makes sense to name it like that.

**Comment by patriknw** [comment]
I fixed the typos

**Comment by patriknw** [comment]
Refs #6577

**Comment by patriknw** [comment]
Refs #16407

**Pull request merged**